### ISSUE TITLE
M207 problems

### ISSUE THREAD

**Issue by y00rek** [dc42/RepRapFirmware#43]
I believe there is still a problem with M207. When I had everything properly calibrated with M207 turned off the first layer height was being set correctly after G32. As soon as I enabled firmware retraction (M207) the first layer is very thin. Either I am doing something wrong or there is a bug there somewhere.

BTW how does one set the IR probe trigger height? Before calibration or after calibration? I do it after running G32. I lower the nozzle until it just grips the paper. Then -
G92 Z0
G1 Z5
G30 S-1
G1 Z5
G30 S-1
G1 Z5
G30 S-1
Then I average out the results and enter it in config.g G31. I restart the printer and run G32 again.
I also run G32 (6 factor, 13 points) before every print.

Here is my config.g
`; $Id: config.g 243 2016-03-29 12:16:59Z yoorek $
;
; Configuration file for yOOrek's Delta printer
;
; Communication and general
;
M111 S0                                         ; Debug off
M550 PyOOrek's Delta                        ; Machine name and Netbios name (can be anything you like)
M551 Preprap                                ; Machine password (used for FTP)
;
; Network
;
M540 P0xBE:0xEF:0xDE:0xAD:0xFE:0xED ; MAC address
M552 P10.0.1.17                 ; IP address (0 = use DHCP)
M554 P10.0.1.1                  ; Gateway
M553 P255.255.255.0             ; Netmask
;
M555 P5                                         ; Set output to look like Repetier (Marlin - P2)
M575 P1 B57600 S1                           ; Comms parameters for PanelDue
;
G21                                                 ; Work in millimetres
G90                                                 ; Send absolute coordinates...
M83                                             ; ...but relative extruder moves
;
; Axis and motor configuration
;
M569 P0 S1                      ; Drive 0 goes forwards (X)
M569 P1 S1                      ; Drive 1 goes forwards (Y)
M569 P2 S1                      ; Drive 2 goes forwards (Z)
M569 P3 S0                      ; Drive 3 goes forwards (extruder 0)
M569 P4 S0                      ; Drive 4 goes forwards (extruder 1)
M574 X2 Y2 Z2 S1                    ; set endstop configuration (all endstops at high end, active high)
;
;**\* The homed height is deliberately set too high in the following - you will adjust it during calibration
;
M665 R160.30 L370.00 B148.00 H431.62 X-0.13 Y0.26 Z0.00 ; set delta radius, diagonal rod length, printable radius and homed height
M666 X-0.19 Y-0.30 Z0.49            ; put your endstop adjustments here, or let auto calibration find them
M92 X200 Y200 Z200              ; Set axis steps/mm
M906 X1650 Y1650 Z1650 E1200 I60    ; Set motor currents (mA) and increase idle current to 60%
M201 X2500 Y2500 Z2500 E2500        ; Accelerations (mm/s^2)
M203 X18000 Y18000 Z18000 E18000    ; Maximum speeds (mm/min)
M566 X500 Y500 Z500 E3000           ; Maximum instant speed changes mm/minute (jerk speeds)
;
; Thermistors
;**\* If you have a Duet board stickered "4.7K", change R1000 to R4700 to the following M305 commands
;
; P0 - bed thermistor
;
M305 P0 T100000 B3950 R4700 H22 L0  ; Put your own H and/or L values here to set the bed thermistor ADC correction
;
; P1 - first nozzle thermistor (E3D V6). David recommends B4388 instead of B4267 stated by e3d people.
;
M305 P1 T100000 B4267 R4700 H22 L76 ; Put your own H and/or L values here to set the first nozzle thermistor ADC correction
;
; P2 - second nozzle thermistor
;
;M305 P2 T100000 B3974 R4700 H30 L0 ; Put your own H and/or L values here to set the second nozzle thermistor ADC correction
;
M570 S180                       ; After a heater has been switched on, wait 180 sec for it to get close to set temperature.
                                ; If it takes longer than this, flag a heater fault.

M143 S275                       ; Set the maximum temp. of the hot-end to 275°C to prevent damage
;
; Tool definitions
;
M563 P0 D0 H1                               ; Define tool 0 (first extruder & hot end)
G10 P0 S0 R50                               ; Set tool 0 operating(S) and standby(R) temperatures
                                ; P - tool number
                                ; X - X offset
                                ; Y - Y offset
                                ; S - active temperature
                                ; R - standby temperature
;
; If you have a dual-nozzle build, un-comment the next 2 lines
;
;M563 P1 D1 H2                              ; Define tool 1
;G10 P1 S0 R50                              ; Set tool 1 operating and standby temperatures
;
M92 E465.11:465.11                      ; Set extruder steps per mm (for two BondTechs)
;
;M207 S8.0 R0.000 F4500 Z0.500      ; Hardware retraction
                                ; S - positive length to retract in mm
                                ; R - positive additional length to un-retract, in mm
                                ; F - feedrate in mm/min
                                ; Z - additional z-lift/hop
;
M572 P0 S0.10                   ; Set tool 1 elasticity compensation (in seconds)
;M572 P1 S0.10                  ; Set tool 2 elasticity compensation (in seconds)
;
; Z probe and compensation definition
;
; If you have an IR z-probe instead of a switch, change P4 to P1 in the following M558 command
;
M558 P1 X0 Y0 Z0 H3 F450 T8000      ; Z-probe is a switch and is not used for homing any axes
                                ; Fnnn - Feed rate (probing speed)
                                                ; Tnnn - Travel speed to probe points

G31 X0 Y0 Z1.28 P500                ; Set the z-probe height and threshold
                                ; If the first layer is too thin then decrease the Z value
                                ; If the first layer is too high then increase the Z value
;
;**\* If you are using axis compensation, put the figures in the following command
;
;M556 S78 X0 Y0 Z0              ; Axis compensation here
M404 N1.75 D0.4                 ; Set filament diameter and nozzle size (machine default, gcode should adjust it)
M208 S1 Z-0.2                   ; set minimum Z
T0                              ; select first hot end`

**Comment by dc42** [comment]
Thank you for your report. Please can you check and report the following:

1. Does it make any difference if you turn off Z hop in the M207 settings?

2. Are you using absolute or relative extruder coordinates in your gcode file?

3. Are you using pressure advance?

4. When you print the same model sliced with and without firmware retraction, is the Z height displayed while printing the first layer correct in both cases?

**Comment by y00rek** [comment]
Sorry for small delay with the answer. Here is what I found (using 1.11-beta1 (2016-04-10))

1) Have not checked this one yet.
2) I'm using relative extruder coordinates
3) No, I'm not using pressure advance in slic3r (it is set to 0)
4) The height is displayed correctly when printing a model with no firmware retraction (Z = 0.10). Same model sliced with firmware retraction displays Z = -0.20 and the first layer is very thin.
5) I mean the layer height.

**Comment by dc42** [comment]
What I think might be happening is that you are using Z hop (nonzero Z parameter in your M207 command) and you have a G11 un-retraction command in the start gcode. Please examine your gcode file to determine whether the first G11 code occurs before or after the first G10 code.

When I asked about pressure advance, I meant RepRapFirmware pressure advance, configured via M572.

**Comment by y00rek** [comment]
Yes I'm using Z hop (M207 S8.0 R0.000 F4500 Z0.500).
I have pressure advance configured (M572 P0 S0.10)
In my code I have G10 first, then G11 (this is not in my start code however). Anyway here is the beginning of my code up until first skirt

```
; generated by Slic3r 1.2.9 on 2016-04-12 at 17:12:07

; Default print settings for my Delta

; external perimeters extrusion width = 0.40mm
; perimeters extrusion width = 0.40mm
; infill extrusion width = 0.40mm
; solid infill extrusion width = 0.40mm
; top infill extrusion width = 0.40mm

M107 ; disable fan
;
; Custom start G-code (from slic3r)
;
M404 N1.75 D0.4 ; Set filament size and nozzle diameter
M300 S5000 P300 ; sound notification
T0 ; Select first extruder
M107 ; disable fan (this is deprecated in RepRap Firmware)
M106 P0 S0 ; disable cooling fan (first extruder)
M106 P1 S0 ; disable cooling fan (second extruder)
;
M190 S70 ; set bed temperature for first layer and wait until reached
M300 S5000 P300 ; sound notification
M104 S210 ; set extruder temperature for first layer and proceed
;
G32 ; Autocalibrate (do it after the bed has reached its temperature)
;
;
; Put a code to make the head go to its waiting position
G0 X-148.00 Y0.00 F8000
;
M109 S210 ; set extruder temperature for first layer and wait until reached
M300 S5000 P1000 ; sound notification
;
; Ready to print
;
G21 ; set units to millimetres
G90 ; use absolute coordinates
M83 ; use relative distances for extrusion
;
; Start of layer 0
;
M300 S5000 P300 ; sound notification
G10 ; retract
G1 Z0.100 F15000.000 ; move to next layer (0)
G1 X-20.125 Y29.043 F15000.000 ; move to first skirt point
G11 ; unretract
G1 X-24.928 Y29.043 E0.13149 F2586.676 ; skirt
```

**Comment by dc42** [comment]
I can see the problem. It's in this bit of the gcode:

G10 ; retract
G1 Z0.100 F15000.000 ; move to next layer (0)
G1 X-20.125 Y29.043 F15000.000 ; move to first skirt point
G11 ; unretract

There is a G1 Z0.100 after the G10 command to set the first layer height, but then the G11 undoes the z-hop, putting the head too low. Looks like your slicer doesn't allow for the possibility of your firmware retraction having a z hop. 

I guess the firmware could be changed to make the un-retraction conditional, but I'm not sure this could be made foolproof, for example if the print is paused and restarted. IMO it should be fixed in the slicer. What slicer are you using? I generated a file using slic3r 1.2.9 with firmware retraction enabled, and it generated correct code:

G1 Z0.200 F7800.000
G10 ; retract
G1 X32.408 Y1.281 F7800.000
G11 ; unretract
G1 X32.066 Y9.944 E0.25743 F1800.000

In this case it moves to the first layer height before retracting instead of after, so the un-retraction doesn't cause an incorrect Z.

**Comment by y00rek** [comment]
I see. I'm using Slic3r 1.2.9 on a Mac. What are your settings in Slic3r? Mine are -
<img width="484" alt="screenshot 2016-04-13 02 06 37" src="https://cloud.githubusercontent.com/assets/799238/14479200/79c80824-011c-11e6-9386-5329d67d477e.PNG">
<img width="382" alt="screenshot 2016-04-13 02 06 46" src="https://cloud.githubusercontent.com/assets/799238/14479201/79e8f2d2-011c-11e6-9c02-5b322933c5e4.PNG">

I will try to play with the settings to see if I can force Slic3r to generate g-code like you.

**Comment by y00rek** [comment]
I was able to generate the first layer code like yours by turning off 'retract on layer change'. Do you have it turned off in yours?

```
;
; Start of layer 0
;
M300 S5000 P300 ; sound notification
G1 Z0.200 F15000.000 ; move to next layer (0)
G10 ; retract
G1 X17.000 Y-13.007 F15000.000 ; move to first skirt point
G11 ; unretract
G1 X18.791 Y-12.611 E0.09584 F5114.789 ; skirt
```

**Comment by dc42** [comment]
No I don't have retract on layer change ticked.

I'll give some thought to tracking whether an un-retraction is pending and adjusting the Z height in any G1 commands when it is. The main issue is likely to be getting it to work properly when a print is paused and then resumed.

**Comment by y00rek** [comment]
I'll leave the _retract on layer change_ turned off for the time being. I guess this is Slic3r's fault for generating g-code in this fashion.

**Comment by dc42** [comment]
I've put a fix for this in firmware 1.17RC3 which I intend to release later today or tomorrow. Please test it with retract on layer change enabled.

**Comment by dc42** [comment]
No feedback received so I am closing this.
